File: app/src/main/java/com/checkin/app/checkin/manager/fragments/ManagerStatsFragment.java
package com.checkin.app.checkin.manager.fragments;

import android.os.Bundle;
import android.view.View;
import android.widget.TextView;

import com.checkin.app.checkin.R;
import com.checkin.app.checkin.data.resource.Resource;
import com.checkin.app.checkin.manager.adapters.ManagerStatsOrderAdapter;
import com.checkin.app.checkin.manager.models.ManagerStatsModel;
import com.checkin.app.checkin.manager.viewmodels.ManagerWorkViewModel;
import com.checkin.app.checkin.misc.BlockingNetworkViewModel;
import com.checkin.app.checkin.misc.fragments.BaseFragment;
import com.checkin.app.checkin.utility.Utils;

import java.util.Locale;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import butterknife.BindView;

public class ManagerStatsFragment extends BaseFragment {

    @BindView(R.id.rv_manager_stats_trending_orders)
    RecyclerView rvTrendingOrders;
    @BindView(R.id.tv_manager_stats_revenue_day)
    TextView tvDayRevenue;
    @BindView(R.id.tv_manager_stats_revenue_week)
    TextView tvWeekRevenue;
    @BindView(R.id.tv_manager_stats_orders_day)
    TextView tvDayOrders;
    @BindView(R.id.tv_manager_stats_orders_week)
    TextView tvWeekOrders;
    @BindView(R.id.tv_manager_stats_session_time)
    TextView tvSessionTime;
    @BindView(R.id.tv_manager_stats_serving_time)
    TextView tvServingTime;

    private ManagerStatsOrderAdapter mAdapter;
    private ManagerWorkViewModel mViewModel;
    private BlockingNetworkViewModel networkViewModel;

    public ManagerStatsFragment() {
    }

    public static ManagerStatsFragment newInstance() {
        return new ManagerStatsFragment();
    }

    @Override
    protected int getRootLayout() {
        return R.layout.fragment_shop_manager_statistics;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        initRefreshScreen(R.id.sr_manager_stats);
        mAdapter = new ManagerStatsOrderAdapter();
        rvTrendingOrders.setLayoutManager(new LinearLayoutManager(getContext(), RecyclerView.VERTICAL, false));
        rvTrendingOrders.setAdapter(mAdapter);

        mViewModel = ViewModelProviders.of(requireActivity()).get(ManagerWorkViewModel.class);
        networkViewModel = ViewModelProviders.of(requireActivity()).get(BlockingNetworkViewModel.class);
        mViewModel.fetchStatistics();
        mViewModel.getRestaurantStatistics().observe(this, input -> {
            if (input == null)
                return;
            handleLoadingRefresh(input);
            networkViewModel.updateStatus(input, null);
            if (input.getStatus() == Resource.Status.SUCCESS && input.getData() != null) {
                setupData(input.getData());
            }

        });

        networkViewModel.shouldTryAgain(s -> {
            mViewModel.fetchStatistics();
            return null;
        });

    }

    private void setupData(@NonNull ManagerStatsModel data) {
        tvDayOrders.setText(String.format(Locale.ENGLISH, "%s orders", data.getDayOrdersCount()));
        tvWeekOrders.setText(String.format(Locale.ENGLISH, "%s orders", data.getWeekOrdersCount()));
        tvDayRevenue.setText(Utils.formatCurrencyAmount(getContext(), data.getDayRevenue()));
        tvWeekRevenue.setText(Utils.formatCurrencyAmount(getContext(), data.getWeekRevenue()));
        tvSessionTime.setText(data.formatAvgSessionTime());
        tvServingTime.setText(data.formatAvgServingTime());

        mAdapter.setData(data.getTrendingOrders());
    }
}

File: app/src/main/java/com/checkin/app/checkin/manager/fragments/ManagerStatsFragment.kt
package com.checkin.app.checkin.manager.fragments

import android.os.Bundle
import android.view.View
import android.widget.TextView
import androidx.fragment.app.activityViewModels
import androidx.lifecycle.Observer
import butterknife.BindView
import com.airbnb.epoxy.EpoxyRecyclerView
import com.checkin.app.checkin.R
import com.checkin.app.checkin.data.resource.Resource
import com.checkin.app.checkin.manager.holders.statsMenuAnalysisGroupRevenue
import com.checkin.app.checkin.manager.holders.statsMenuAnalysisItemRevenue
import com.checkin.app.checkin.manager.models.ManagerStatsModel
import com.checkin.app.checkin.manager.viewmodels.ManagerWorkViewModel
import com.checkin.app.checkin.misc.fragments.BaseFragment
import com.checkin.app.checkin.misc.holders.textSectionModelHolder
import com.checkin.app.checkin.utility.Utils
import com.checkin.app.checkin.utility.isNotEmpty

class ManagerStatsFragment : BaseFragment() {
    override val rootLayout: Int = R.layout.fragment_shop_manager_statistics

    @BindView(R.id.epoxy_rv_manager_stats_menu_analysis)
    internal lateinit var epoxyRvMenuAnalysis: EpoxyRecyclerView

    @BindView(R.id.tv_manager_stats_revenue_live)
    internal lateinit var tvLiveCartRevenue: TextView

    @BindView(R.id.tv_manager_stats_revenue_yesterday)
    internal lateinit var tvYesterdayRevenue: TextView

    @BindView(R.id.tv_manager_stats_revenue_today)
    internal lateinit var tvDayRevenue: TextView

    @BindView(R.id.tv_manager_stats_revenue_week)
    internal lateinit var tvWeekRevenue: TextView

    @BindView(R.id.tv_manager_stats_orders_live)
    internal lateinit var tvLiveCartOrders: TextView

    @BindView(R.id.tv_manager_stats_orders_yesterday)
    internal lateinit var tvYesterdayOrders: TextView

    @BindView(R.id.tv_manager_stats_orders_today)
    internal lateinit var tvDayOrders: TextView

    @BindView(R.id.tv_manager_stats_orders_week)
    internal lateinit var tvWeekOrders: TextView

    @BindView(R.id.tv_manager_analysis_acceptance_time)
    internal lateinit var tvOrderAcceptanceTime: TextView

    @BindView(R.id.tv_manager_analysis_session_time)
    internal lateinit var tvSessionTime: TextView

    @BindView(R.id.tv_manager_analysis_serving_time)
    internal lateinit var tvServingTime: TextView

    private val mViewModel: ManagerWorkViewModel by activityViewModels()

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        initRefreshScreen(R.id.sr_manager_stats)
        epoxyRvMenuAnalysis.withModels {
            mViewModel.restaurantStatistics.value?.data?.trendingItems.takeIf { it.isNotEmpty() }?.also {
                textSectionModelHolder {
                    withRedSmallLayout()
                    id("trending_items")
                    heading("TRENDING ITEMS IN LAST 30 DAYS")
                }
                it.forEachIndexed { idx, data ->
                    statsMenuAnalysisItemRevenue {
                        id(idx)
                        data(data)
                    }
                }
            }

            mViewModel.restaurantStatistics.value?.data?.poorItems.takeIf { it.isNotEmpty() }?.also {
                textSectionModelHolder {
                    withRedSmallLayout()
                    id("poor_items")
                    heading("LEAST SELLING ITEM IN LAST 30 DAYS")
                }
                it.forEachIndexed { idx, data ->
                    statsMenuAnalysisItemRevenue {
                        id(idx)
                        data(data)
                    }
                }
            }

            mViewModel.restaurantStatistics.value?.data?.trendingGroups.takeIf { it.isNotEmpty() }?.also {
                textSectionModelHolder {
                    withRedSmallLayout()
                    id("trending_groups")
                    heading("TOP GROUP IN LAST 30 DAYS")
                }
                it.forEachIndexed { idx, data ->
                    statsMenuAnalysisGroupRevenue {
                        id(idx)
                        index(idx + 1)
                        data(data)
                    }
                }
            }
        }

        mViewModel.fetchStatistics()
        mViewModel.restaurantStatistics.observe(viewLifecycleOwner, Observer {
            it?.let { input ->
                handleLoadingRefresh(input)
                if (input.status === Resource.Status.SUCCESS && input.data != null) {
                    setupData(input.data)
                }
            }
        })
    }

    private fun setupData(data: ManagerStatsModel) {
        tvLiveCartOrders.text = "%.0f Orders".format(data.countOrders.live)
        tvDayOrders.text = "%.0f Orders".format(data.countOrders.day)
        tvYesterdayOrders.text = "%.0f Orders".format(data.countOrders.yesterday)
        tvWeekOrders.text = "%.0f Orders".format(data.countOrders.week)
        tvLiveCartRevenue.text = Utils.formatCurrencyAmount(context, data.revenue.live)
        tvDayRevenue.text = Utils.formatCurrencyAmount(context, data.revenue.day)
        tvYesterdayRevenue.text = Utils.formatCurrencyAmount(context, data.revenue.yesterday)
        tvWeekRevenue.text = Utils.formatCurrencyAmount(context, data.revenue.week)

        tvSessionTime.text = "%s Session time".format(data.formatAvgSessionTime())
        tvServingTime.text = "%s Serving time".format(data.formatAvgServingTime())
        tvOrderAcceptanceTime.text = "%s Order Accepting time".format(data.formatOrderAcceptingTime())
    }

    override fun updateScreen() {
        super.updateScreen()
        mViewModel.fetchStatistics()
    }

    companion object {
        fun newInstance(): ManagerStatsFragment = ManagerStatsFragment()
    }
}